### DISCUSSION TITLE
Keycloak extension dev: how to enable support for @Valid annotation?

### DISCUSSION THREAD

**Discussion by raoulvdberge** [keycloak/keycloak#22687] [Q&A]
Hello,
What is the intended way to validate the request body of a `RealmResourceProvider` resource? 
Keycloak validation API is on the classpath through `keycloak-model-jpa`, but it appears like there's no implementation configured (like hibernate validation).
I've then tried to add hibernate validation API to my classpath but it seems like it's not being used in the JAX-RS context.
Any pointers are welcome.

**Comment by alphana** [comment]
Dear team, we have exactly same problem while handling jakarta validation results. 
Extending with an alternative, is there any way to provide/register a custom javax.ws.rs.ext.ExceptionMapper to the RealmResource?

**Comment by raoulvdberge** [comment]
I have solved this issue on my end by calling validate on the Hibernate Validator manually instead of using the `@Valid` annotation.

**Reply by milic-marko** [reply]
@raoulvdberge , Thanks for the hint, but how did you solve the problem of 

`jakarta.validation.NoProviderFoundException: Unable to create a Configuration, because no Jakarta Bean Validation provider could be found. Add a provider like Hibernate Validator (RI) to your classpath.`

As wehen I try to call Hibernate Validator manually I get the error above.

**Reply by raoulvdberge** [reply]
Make sure to add `hibernate-validator` to your classpath. You can do this by adding it to the `providers` folder.
In Docker:
```COPY --chown=keycloak:keycloak /target/dependency/hibernate-validator.jar /opt/keycloak/providers/``` 

You could use the `maven-dependency-plugin` to extract this jar:
```xml
            <plugin>
                <groupId>org.apache.maven.plugins</groupId>
                <artifactId>maven-dependency-plugin</artifactId>
                <executions>
                    <execution>
                        <id>unpack</id>
                        <phase>package</phase>
                        <goals>
                            <goal>copy</goal>
                        </goals>
                        <configuration>
                            <artifactItems>
                                <artifact>
                                    <groupId>org.hibernate.validator</groupId>
                                    <artifactId>hibernate-validator</artifactId>
                                    <type>jar</type>
                                    <destFileName>hibernate-validator.jar</destFileName>
                                </artifact>
                            </artifactItems>
                        </configuration>
                    </execution>
                </executions>
            </plugin>
``` 

Then, you should be able to obtain a `Validator` instance:
```java
    private static final Validator VALIDATOR = jakarta.validation.Validation.byDefaultProvider()
        .configure()
        .messageInterpolator(new ParameterMessageInterpolator())
        .buildValidatorFactory()
        .getValidator();

    public static <T> void validate(final T subject) {
        final Set<ConstraintViolation<Object>> violations = VALIDATOR.validate(subject);
        if (!violations.isEmpty()) {
            throw new InvalidOperationException(violations);
        }
    }
```
Hope this helps.

**Reply by milic-marko** [reply]
@raoulvdberge , Thank you a lot! I was wondering is there any convinient way to intercept the exception in keycloak and modify response, something similar to @ControllerAdvice in Spring?

**Reply by danielFesenmeyer** [reply]
Afaik, there is no resource-specific exception handling mechanism for Keycloak/Quarkus.

I currently see two options:

### Option 1: Custom exception handling with lambda

What I did for such cases: I've created a method for exception handling like this:
`<T> T runWithExceptionHandling(final Supplier<T> supplier)`

The method has try/catch logic (with logging), mapping domain-specific exceptions to Jakarta-RS exceptions like `jakarta.ws.rs.BadRequestException`. This way, [KeycloakErrorHandler](https://github.com/keycloak/keycloak/blob/main/services/src/main/java/org/keycloak/services/error/KeycloakErrorHandler.java) will return the excepted status and body.

Every endpoint logic is passed to the exception handling method.

### Option 2: Global exception mapper

Implement your own global `ExceptionMapper`, like it's done for [KeycloakErrorHandler](https://github.com/keycloak/keycloak/blob/main/services/src/main/java/org/keycloak/services/error/KeycloakErrorHandler.java). Maybe you have to use the `@Priority` annotation to make sure your mapper takes precedence over `KeycloakErrorHandler`.

**Comment by shawkins** [comment]
For keycloak 24+ I believe you need quarkus-hibernate-validator and the missing transitive dependencies, for KC 24 the provider directory needs:

io.quarkus:quarkus-hibernate-validator:3.8.5
org.glassfish.expressly:expressly:5.0.0
org.hibernate.validator:hibernate-validator:8.0.1.Final
io.smallrye.config:smallrye-config-validator:3.5.4

**Reply by shawkins** [reply]
Unfortunately this does not work - it is also missing the quarkus-hibernate-validator-deployment dependnecy. However just including that in the provider directory does not influence the lib/lib/deployment deployment class path.

@geoand is there a supported way to add a quarkus extension after the fact to keycloak, which is built as a mutable jar?

**Reply by geoand** [reply]
No, there is not

**Reply by shawkins** [reply]
Thanks @geoand for the confirmation. 

So a motivated community user could add the quarkus-hibnerate-validator dependency to our quarkus/runtime pom.xml, and the  quarkus-hibnerate-validator-deployment dependency to our quarkus/deployment pom.xml and  build a customized distribution, but otherwise will need to use workaround of manual validation outlined above for now.

**Comment by banks047** [comment]
This still seems to be broken, are there any updates regarding @Valid annotations no longer working in latest Keycloak builds or is the manual validation the only way from now on?

**Reply by shawkins** [reply]
@banks047 see https://github.com/keycloak/keycloak/discussions/22687#discussioncomment-10128585 - related to this there is also now https://github.com/mabartos/keycloak-quarkus-extensions
